File: scripts/verificar_aislamiento_datos.py
import os
import sys
from dotenv import load_dotenv
from supabase import create_client, Client

# Cargar variables de entorno
load_dotenv()

SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_SERVICE_ROLE_KEY = os.getenv("SUPABASE_SERVICE_ROLE")  # Usar service role para inspección
SUPABASE_ANON_KEY = os.getenv("SUPABASE_ANON_KEY")

if not SUPABASE_URL or not SUPABASE_SERVICE_ROLE_KEY:
    print("Error: Faltan variables de entorno SUPABASE_URL o SUPABASE_SERVICE_ROLE")
    sys.exit(1)

def verificar_rls():
    """ Verifica que RLS esté habilitado en todas las tablas publicas. """
    print("\n--- Verificando RLS en Tablas ---")
    
    # Cliente con permisos administrativos para inspeccionar
    supabase_ad: Client = create_client(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY)
    
    # Consulta SQL para obtener tablas y estado de RLS
    # Nota: Esto requiere que el rol de servicio tenga permisos de ejecución SQL o consultar pg_catalog
    # Como fallback, intentaremos usar la API REST si SQL no está disponible directamente via py-supabase (que suele ser solo para data)
    # PERO, supabase-py no tiene "execute_sql" nativo fácil. 
    # Usaremos una aproximación: intentar acceder a tablas conocidas sin autenticación y esperar fallo.
    
    # Lista de tablas críticas conocidas (Ajustar según esquema real)
    tablas_criticas = ["tasks", "profiles", "inbox_items", "projects"]
    veredicto_final = True

    # 1. Prueba de Acceso Anónimo (Debe fallar o dar 0 resultados si RLS funciona y no hay policy public)
    print("1. Probando acceso anónimo (debe ser denegado o vacío)...")
    supabase_anon: Client = create_client(SUPABASE_URL, SUPABASE_ANON_KEY)
    
    for tabla in tablas_criticas:
        try:
            # Intentar leer todo
            res = supabase_anon.table(tabla).select("*").limit(1).execute()
            # Si devuelve datos, ES UN RIESGO (a menos que sea intencional pública)
            if res.data and len(res.data) > 0:
                print(f"ALERTA: Tabla '{tabla}' devuelve datos a usuario anónimo! POSIBLE FUGA RLS.")
                veredicto_final = False
            else:
                print(f"OK: Tabla '{tabla}' segura para anónimos (0 registros o error).")
        except Exception as e:
            # Si da error de permiso, es bueno en este contexto
            print(f"OK: Tabla '{tabla}' acceso denegado (Excepción capturada: {str(e)}).")

    return veredicto_final

if __name__ == "__main__":
    if verificar_rls():
        print("\n[PASSED] Verificación de Seguridad RLS completada satisfactoriamente.")
        sys.exit(0)
    else:
        print("\n[FAILED] Se detectaron problemas de seguridad RLS.")
        sys.exit(1)

File: scripts/verificar_aislamiento_datos.js
import { createClient } from '@supabase/supabase-js';
import dotenv from 'dotenv';
import path from 'path';
import { fileURLToPath } from 'url';

// Configuración de __dirname para ES Modules
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Cargar variables de entorno desde el root
dotenv.config({ path: path.resolve(__dirname, '../.env') });

const SUPABASE_URL = process.env.VITE_SUPABASE_URL || process.env.SUPABASE_URL;
const SUPABASE_ANON_KEY = process.env.VITE_SUPABASE_ANON_KEY || process.env.SUPABASE_ANON_KEY;

if (!SUPABASE_URL || !SUPABASE_ANON_KEY) {
    console.error("Error: Faltan variables de entorno VITE_SUPABASE_URL o VITE_SUPABASE_ANON_KEY");
    process.exit(1);
}

async function verificarRLS() {
    console.log("\n--- Verificando RLS en Tablas ---");

    const tablasCriticas = ["tasks", "profiles", "inbox_items", "projects"];
    let veredictoFinal = true;

    console.log("1. Probando acceso anónimo (debe ser denegado o vacío)...");

    // Cliente anónimo para simular atacante externo
    const supabaseAnon = createClient(SUPABASE_URL, SUPABASE_ANON_KEY);

    for (const tabla of tablasCriticas) {
        try {
            const { data, error } = await supabaseAnon.from(tabla).select("*").limit(1);

            if (error) {
                // Si da error de permiso, es bueno en este contexto
                console.log(`OK: Tabla '${tabla}' acceso denegado o error (Esperado: ${error.message}).`);
            } else if (data && data.length > 0) {
                console.error(`ALERTA: Tabla '${tabla}' devuelve datos a usuario anónimo! POSIBLE FUGA RLS.`);
                veredictoFinal = false;
            } else {
                console.log(`OK: Tabla '${tabla}' segura para anónimos (0 registros).`);
            }
        } catch (e) {
            console.log(`OK: Tabla '${tabla}' generó excepción: ${e.message}`);
        }
    }

    return veredictoFinal;
}

// Ejecutar
verificarRLS().then((success) => {
    if (success) {
        console.log("\n[PASSED] Verificación de Seguridad RLS completada satisfactoriamente.");
        process.exit(0);
    } else {
        console.log("\n[FAILED] Se detectaron problemas de seguridad RLS.");
        process.exit(1);
    }
}).catch(err => {
    console.error("Error inesperado:", err);
    process.exit(1);
});
